### PR TITLE
fix register type race

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Actually seen in recursor, but auth could also have it in theory.

The issue is that you cannot call type register functions while the rest of the threads are doing their work. The actual fix is in `reczones.cc`, but also guard against this happening in other places (for both rec and auth), and while there remove the unused `unregist` function.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

**Review by rgacogne — COMMENTED** [review]
I see that `reportBasicTypes()` is only used by unit tests, `reportOtherTypes()` is no longer used after this PR, so I think they could be removed in favor of always calling `reportAllTypes()`. Once this is done I don't think anyone should call `::report()` directly (the remote backend unit tests currently do, but it looks like it could call `reportAllTypes() instead`), so perhaps it would make sense to disallow that (for example by making it take a required parameter to a const-reference to an opaque object, that can only be constructed by `reportAllTypes()`). This way we can call `DNSRecordContent::lock()` from `reportAllTypes()` directly, preventing a mistake later where we call `reportAllTypes()` without calling `lock()`.
Does that make sense to you? If so I can look into it if you want me to.

File: pdns/dnsparser.hh [hunk 1]
```diff
@@ -293,6 +292,7 @@ protected:
   static t2namemap_t& getT2Namemap();
   static n2typemap_t& getN2Typemap();
   static zmakermap_t& getZmakermap();
+  static bool d_locked;
```
[the remark this answers is not in the document]
Comment (rgacogne) [reply pdns/dnsparser.hh:295]: Alright, I'll have a look!

**Comment by rgacogne** [comment]
I pushed a commit implementing my proposal, please let me know if it looks good to you.

**Comment by omoerbeek** [comment]
I'm mostly happy with this. One remark: we tend to move to  `/* unsued */` instead of a `void` cast of the unused parameter. If you agree I can change that before merge.

**Comment by rgacogne** [comment]
> I'm mostly happy with this. One remark: we tend to move to `/* unsued */` instead of a `void` cast of the unused parameter. If you agree I can change that before merge.

Sure!

File: pdns/auth-main.cc [hunk 2]
```diff
@@ -1210,6 +1210,7 @@ int main(int argc, char** argv)
 {
   versionSetProduct(ProductAuthoritative);
   reportAllTypes(); // init MOADNSParser
+  DNSRecordContent::lock();
```
Comment (rgacogne) [inline pdns/auth-main.cc:1213]: I forgot to remove this call.

File: pdns/recursordist/rec-main.cc [hunk 3]
```diff
@@ -3153,8 +3153,8 @@ int main(int argc, char** argv)
   g_argc = argc;
   g_argv = argv;
   versionSetProduct(ProductRecursor);
-  reportBasicTypes();
-  reportOtherTypes();
+  reportAllTypes();
+  DNSRecordContent::lock();
```
Comment (rgacogne) [inline pdns/recursordist/rec-main.cc:3157]: And this one as well.

Comment (omoerbeek) [reply pdns/recursordist/rec-main.cc:3157]: I'll take care

File: pdns/dnsrecords.hh [hunk 4]
```diff
@@ -1008,15 +1009,11 @@ void RNAME##RecordContent::toPacket(DNSPacketWriter& pw) const
...
-{                                                                                                  \
-  regist(1, QType::RNAME, &RNAME##RecordContent::make, &RNAME##RecordContent::make, #RNAME);              \
-  regist(254, QType::RNAME, &RNAME##RecordContent::make, &RNAME##RecordContent::make, #RNAME);            \
-}                                                                                                  \
-void RNAME##RecordContent::unreport(void)                                                          \
+void RNAME##RecordContent::report(const ReportIsOnlyCallableByReportAllTypes& guard)               \
 {                                                                                                  \
-  unregist(1, QType::RNAME);                                                                              \
-  unregist(254, QType::RNAME);                                                                            \
+  (void)guard;                                                                                     \
```
Comment (rgacogne) [inline pdns/dnsrecords.hh:1014]: You might want to change this one as well for consistency.

**Review by rgacogne — APPROVED** [review]

**Review by rgacogne — APPROVED** [review]

**Pull request merged**